Enable support for Postgres connection external to Docker container

rustprooflabs opened the pull request:
# Details

Work in progress to close #194.

Current notes to run with external Postgres are rough, not fully tested, in `/docs/DOCKER-RUN-EXTERNAL-PG.md`.

The switch in functionality is setting `POSTGRES_HOST` to a valid external Postgres host.  Default is `localhost` for in-Docker operation.

* In-Docker (`localhost`) will use hard coded `pgosm` DB name
* Setting `POSTGRES_DB` when `POSTGRES_HOST` is `localhost` does nothing
* External Postgres will always skip the db prep steps (drop/create)

Running `make` passes everything so I'm relatively confident I haven't broken anything obvious. A quick test with an external host showed it attempted to connect but failed b/c I hadn't setup the target DB.  Requires more testing, docs and polish.

rustprooflabs commented:
Latest commits add additional safety check against drop/create database in non-Docker environment. Add a check to enforce Postgres 12+ since Postgres version is no longer coupled directly with runtime.  Raises `EnvironmentError` if Postgres 11 or older detected.

:+1: Successfully completed basic testing with PgOSM Flex Docker container running on laptop pushing data into Postgres 12 running on Raspberry Pi 4 on LAN.

:bulb: Docs should probably suggest including `--skip-dump` for a default, seems this use case is like likely to care about that file and the overhead of pg_dump isn't always negligible.

:warning: In current state, Target DB requires `postgres` database for admin connections.  This means it will fail on instances w/out `postgres` db such as AWS RDS, etc.

:warning: Need some tests covering the switch logic RE data prep steps.  Probably needs refactoring to make easier to test.

rustprooflabs commented:
@jacopofar I think the core of this is working and the new docs explain the basic functionality, comments above explain current known caveats. If you happen to have any time to test/review it would be greatly appreciated!

In the meantime I'll keep working away at polishing this up and getting it ready to merge in.

jacopofar commented:
Thanks, I'll try it.
Why is it using different variables instead of a single connection string? I saw this comment

>  PGOSM_CONN is required by Lua scripts for osm2pgsql. This should
>  be the only place a connection string is defined outside of Sqitch usage.

I know the database name is needed as a string to enable the DB creation, but why not extract it from the connection string when needed and pass around only that? It's not a big deal but I find a connections string to be easy to use.

rustprooflabs commented:
Using the individual connection string components is a side effect of the base PostGIS (and Postgres) Docker images.  To run the container you must define `POSTGRES_PASSWORD` or get this error:

```
Error: Database is uninitialized and superuser password is not specified.
       You must specify POSTGRES_PASSWORD to a non-empty value for the
       superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".
```

Because of this requirement I decided to stick to the individual connection string parts to make the usage through Docker consistent.  Also worth noting is that the environment vars for [Postgres in Docker](https://registry.hub.docker.com/_/postgres/) is not the same as what [Postgres itself uses](https://www.postgresql.org/docs/current/libpq-envars.html) for standard environment variables.  The Docker image uses `POSTGRES_PASSWORD` but Postgres defines `PGPASSWORD`.

jacopofar commented:
Ah fine then, I didn't think about that

rustprooflabs commented:
Fixed the pre-Pg12 version check to work as expected. To test change the Dockerfile image to 11 instead of 14, build and try running. 

Other improvements:

* External Postgres connections do not expect a `postgres` database to exist
* Added a few tests covering related functionality
* Updated two test runs in Makefile to run slightly faster, leaving one with default switches
* Improvements to error handling, comments, log messages
